src/blockstore/reader.zig

```diff
@@ -0,0 +1,2037 @@
+                                all_ranges_start_index,
+                                all_ranges_end_index,
+                            },
+                        );
+                        return error.CorruptedBlockstore;
+                    }
+                }
+                self.logger.errf("Missing shred for slot {}, index {}", .{ slot, index });
+                return error.InvalidShredData;
+            }
+        }
+
+        var entries = ArrayList(Entry).init(self.allocator);
+        for (completed_ranges.items) |range| {
+            const start_index, const end_index = range;
+
+            // The indices from completed_ranges refer to shred indices in the
+            // entire block; map those indices to indices within data_shreds
+            const range_start_index: usize = @intCast(start_index - all_ranges_start_index);
+            const range_end_index: usize = @intCast(end_index - all_ranges_start_index);
+            const range_shreds: []DataShred =
+                data_shreds.items[range_start_index .. range_end_index + 1];
+
+            const last_shred = range_shreds[range_shreds.len - 1];
+            std.debug.assert(last_shred.dataComplete() or last_shred.isLastInSlot());
+            // self.logger.tracef("{any} data shreds in last FEC set", data_shreds.items.len);
+
+            const bytes = shredder.deshred(self.allocator, range_shreds) catch |e| {
+                self.logger.errf("failed to deshred entries buffer from shreds: {}", .{e});
+                return e;
+            };
+            defer bytes.deinit();
+            const these_entries = sig.bincode
+                .readFromSlice(self.allocator, []Entry, bytes.items, .{}) catch |e| {
+                self.logger.errf("failed to deserialize entries from shreds: {}", .{e});
+                return e;
+            };
+            defer self.allocator.free(these_entries);
+            try entries.appendSlice(these_entries);
+        }
+        return entries;
+    }
+
+    // pub fn is_last_fec_set_full missing ???
+
+    /// Returns a mapping from each elements of `slots` to a list of the
+    /// element's children slots.
+    ///
+    /// Analogous to [get_slots_since](https://github.com/anza-xyz/agave/blob/15dbe7fb0fc07e11aaad89de1576016412c7eb9e/ledger/src/blockstore.rs#L3821)
+    pub fn getSlotsSince(self: *Self, slots: []const Slot) !AutoHashMap(Slot, ArrayList(Slot)) {
+        // TODO perf: support multi_get in db
+        var map = AutoHashMap(Slot, ArrayList(Slot)).init(self.allocator);
+        errdefer {
+            var iter = map.iterator();
+            while (iter.next()) |entry| {
+                entry.value_ptr.deinit();
+            }
+            map.deinit();
+        }
+        for (slots) |slot| {
+            if (try self.db.get(schema.slot_meta, slot)) |meta| {
+                errdefer meta.next_slots.deinit();
+                var cdi = meta.completed_data_indexes;
+                cdi.deinit();
```

Comment:
why do this?

Reply:
`meta` as a whole cannot be deinitialized because next_slots moving to the map. But when meta goes out of scope, completed_data_indexes is going to leak unless its backing memory is freed. It's assigned to a var because hashmaps need a mutable pointer to deinit.